src/Analyzer/Passes/FunctionToSubcolumnsPass.cpp

```diff
@@ -16,65 +16,239 @@
             return;
 
-        if (first_argument_column_node->getColumnName() == "__grouping_set")
+        auto add_key_columns = [&](const auto & key_columns)
+        {
+            for (const auto & column_name : key_columns)
+            {
+                Identifier identifier({table_name, column_name});
+                data.all_key_columns.insert(identifier);
+            }
+        };
+
+        const auto & metadata_snapshot = table_node.getStorageSnapshot()->metadata;
+
+        const auto & primary_key_columns = metadata_snapshot->getColumnsRequiredForPrimaryKey();
+        add_key_columns(primary_key_columns);
+
+        const auto & partition_key_columns = metadata_snapshot->getColumnsRequiredForPartitionKey();
+        add_key_columns(partition_key_columns);
+
+        for (const auto & index : metadata_snapshot->getSecondaryIndices())
+        {
+            const auto & index_columns = index.expression->getRequiredColumns();
+            add_key_columns(index_columns);
+        }
+    }
+
+    void enterImpl(const ColumnNode & column_node)
+    {
+        if (column_node.getColumnName() == "__grouping_set")
             return;
 
-        auto column_source = first_argument_column_node->getColumnSource();
+        auto column_source = column_node.getColumnSource();
         auto * table_node = column_source->as<TableNode>();
-
         if (!table_node)
             return;
 
-        const auto & storage = table_node->getStorage();
-        if (!storage->supportsSubcolumns())
-            return;
+        auto table_name = table_node->getStorage()->getStorageID().getFullTableName();
+        Identifier qualified_name({table_name, column_node.getColumnName()});
 
-        auto column = first_argument_column_node->getColumn();
+        ++data.indentifiers_count[qualified_name];
+    }
+
+    void enterImpl(const FunctionNode & function_node, const ColumnNode & first_argument_column_node, const TableNode & table_node)
+    {
+        const auto & function_arguments_nodes = function_node.getArguments().getNodes();
+        const auto & function_name = function_node.getFunctionName();
+
+        auto column = first_argument_column_node.getColumn();
         WhichDataType column_type(column.type);
 
+        auto table_name = table_node.getStorage()->getStorageID().getFullTableName();
+        Identifier qualified_name({table_name, column.name});
+
+        if (function_arguments_nodes.size() == 1)
+        {
+            if (column_type.isArray())
+            {
+                if (function_name == "length" || function_name == "empty" || function_name == "notEmpty")
+                    ++data.optimized_identifiers_count[qualified_name];
+            }
+            else if (column_type.isNullable())
+            {
+                if (function_name == "isNull" || function_name == "isNotNull")
+                    ++data.optimized_identifiers_count[qualified_name];
+            }
+            else if (column_type.isMap())
```

Comment:
No, I'll fix it.

Reply:
I added the optimization for `length(map_column)`, but currently not for `length(mapKeys(map_column))` because it's harder to imlement. Maybe I'll add it later in a separate PR.

Reply:
What if length(map.keys) ?

Reply:
`length(map.keys)` is not supported too. It's almost the same case as `length(mapKeys(map_column))` and now it doesn't work because there is just no subcolumn `map_column.keys.size0`, there is only `map_column.size0`. To support it we need to add some hack and handle this case separately.

Reply:
I see, i guessed, that because map.keys is real Array, it just reduces to use case length(array)